x11-packages/xfce4-cpugraph-plugin/0001-cpu-data.patch

```diff
@@ -0,0 +1,80 @@
+diff --git a/panel-plugin/os.cc b/panel-plugin/os.cc
+index a8c9e40..887c5f3 100644
+--- a/panel-plugin/os.cc
++++ b/panel-plugin/os.cc
+@@ -42,6 +42,11 @@
+ 
+ using namespace std;
+ 
++#ifdef __ANDROID__
++#include <math.h>
++#include <unistd.h>
++#endif
++
+ #if defined (__linux__) || defined (__FreeBSD_kernel__)
+ #define PROC_STAT "/proc/stat"
+ #define PROCMAXLNLEN 256 /* should make it */
+@@ -78,7 +83,7 @@ using namespace std;
+ static kstat_ctl_t *kc;
+ #endif
+ 
+-#if defined (__linux__) || defined (__FreeBSD_kernel__)
++#if (defined (__linux__) && !defined (__ANDROID__)) || defined (__FreeBSD_kernel__)
+ void
+ read_cpu_data (unordered_map<guint, CpuData> &data, unordered_map<guint, guint> &cpu_to_index)
+ {
+@@ -401,12 +406,53 @@ read_cpu_data (unordered_map<guint, CpuData> &data)
+     data[0].load /= nb_cpu;
+     return true;
+ }
++#elif defined (__ANDROID__)
++
++static guint
++detect_cpu_number ()
++{
++    return sysconf(_SC_NPROCESSORS_ONLN);
++}
++
++static bool
++read_cpu_data (unordered_map<guint, CpuData> &data)
++{
++    FILE* top_process;
++    int pid;
++    float cpu_usage;
++    int cpu_index;
++    char line[100];
++
++    if (G_UNLIKELY(data.size() == 0))
++        return false;
++
++    const size_t nb_cpu = data.size()-1;
++
++    for (int i = 0; i <= nb_cpu; i++) {
++        data[i].user = 0;
++        data[i].load = 0;
++    }
++
++    if ((top_process = popen("top -H -b -q -o TID,%CPU,CPU -s 2 -n 1", "r"))) {
```

Comment:
This actually only works on Android 9.0 or later, but I don't think there are many users running X11 programs on Android 7 and Android 8, so I think this is acceptable here. 

PS: I'm just commenting here in case other devs port similar logic to some other frequently used tool (e.g. `htop`).

https://android.googlesource.com/platform/external/toybox/+/refs/tags/android-7.0.0_r36/toys/posix/ps.c#124
```
  help
    usage: top [-m] [ -d seconds ] [ -n iterations ]
    Show process activity in real time.
    -k	Fallback sort FIELDS (default -S,-%CPU,-ETIME,-PID)
    -o	Show FIELDS (def PID,USER,PR,NI,VIRT,RES,SHR,S,%CPU,%MEM,TIME+,CMDLINE)
    -s	Sort by field number (1-X, default 9)
```

https://android.googlesource.com/platform/external/toybox/+/refs/tags/android-8.0.0_r51/toys/posix/ps.c
```
  help
    usage: top [-H] [-k FIELD,] [-o FIELD,] [-s SORT]
    Show process activity in real time.
    -H	Show threads
    -k	Fallback sort FIELDS (default -S,-%CPU,-ETIME,-PID)
    -o	Show FIELDS (def PID,USER,PR,NI,VIRT,RES,SHR,S,%CPU,%MEM,TIME+,CMDLINE)
    -O	Add FIELDS (replacing PR,NI,VIRT,RES,SHR,S from default)
    -s	Sort by field number (1-X, default 9)
```

https://android.googlesource.com/platform/external/toybox/+/refs/tags/android-9.0.0_r61/toys/posix/ps.c

```
  help
    usage: top [-Hbq] [-k FIELD,] [-o FIELD,] [-s SORT] [-n NUMBER] [-m LINES] [-d SECONDS] [-p PID,] [-u USER,]
    Show process activity in real time.
    -H	Show threads
    -k	Fallback sort FIELDS (default -S,-%CPU,-ETIME,-PID)
    -o	Show FIELDS (def PID,USER,PR,NI,VIRT,RES,SHR,S,%CPU,%MEM,TIME+,CMDLINE)
    -O	Add FIELDS (replacing PR,NI,VIRT,RES,SHR,S from default)
    -s	Sort by field number (1-X, default 9)
    -b	Batch mode (no tty)
    -d	Delay SECONDS between each cycle (default 3)
    -m	Maximum number of tasks to show
    -n	Exit after NUMBER iterations
    -p	Show these PIDs
    -u	Show these USERs
    -q	Quiet (no header lines)
    Cursor LEFT/RIGHT to change sort, UP/DOWN move list, space to force
    update, R to reverse sort, Q to exit.
```

Reply:
`/proc/stat` is prohibited from reading since Android 9.0. It would be better to dynamically check the api level using `android_get_device_api_level` and select which method to use. If >= 28, use the htop method, otherwise use `/proc/stat` method.

Reply:
Rewrote to use `/proc/#/task/#/stat` and keep the `top` approach in commit history just in case we need in the future.